### PR TITLE
Make system chaining use a normal system

### PR DESCRIPTION
**Pull request by DJMcNab**
# Objective

- The API for our system abstraction is functions, not the `System` trait.
- Demonstrate that this can be used for higher order systems
- Allow it to be used for higher order systems

## Solution

- Change `SystemParamFunction` to allow higher-order systems to be created in terms of functions rather than Systems.
- Migrate `ChainSystem` to this.

Please note, that without GATs, as far as I know, we cannot have `chain` remain as a method. However, GATs are on the path to stabilisation, hopefully in a form sufficiently powerful for this case.

## Open questions

Is `chain` the correct name? I think it's probably a fine name, but the idiomatic style is for functions which return systems to have the `_system` suffix. Do we make an exception for `chain` due to the old name or higher-order nature?

### Changed

The `chain` method on system functions has been removed. Use the `bevy_ecs::system::chain` function instead (which is in the `bevy_ecs` prelude).

## Migration Guide

The `chain` method on system functions has been removed. Use the `bevy_ecs::system::chain` function instead (which is in the `bevy_ecs` prelude). If previously you used chain as:
```rust
do_thing.chain(handler)
```
you should instead use:
```rust
chain(do_thing, handler)
```

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
This is a much simpler API, and should significantly simplify working with system chaining elsewhere.

It makes sense that we can do this now that systems as closures is supported.

**Comment by alice-i-cecile** [comment]
>  nil / nile 🧪: it'd need TAIT, not gat s
> [2:14 PM] nil / nile 🧪: and so yes, personally I don't think it should be merged in yet
> [2:15 PM] nil / nile 🧪: it
> - doesn't work for dynamic systems
> - worsens the userspace ergonomics
> [2:16 PM] nil / nile 🧪: and while cleaner code is always a plus, given that it's a piece of code not much else depends on i don't see the point in merging it yet

From @TheRawMeatball on [Discord](https://discord.com/channels/691052431525675048/749335865876021248/980896995474571294)

In light of that, I think it's reasonable to block this on [the TAIT RFC](https://rust-lang.github.io/rfcs/2515-type_alias_impl_trait.html#summary) and the [associated Rustlang issue](https://github.com/rust-lang/rust/issues/63063).

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
We should swap this back to a method, as GATs have landed 🎉

**Comment by alice-i-cecile** [comment]
This is no longer blocked! Want to rebase, or should I stick an Adopt-Me on this?

**Comment by DJMcNab** [comment]
Can you clarify what changed to unblock it?

**Comment by alice-i-cecile** [comment]
Stageless merged :)

**Comment by DJMcNab** [comment]
I must admit, I'm still not following how that solves the issue that we can't add the chain method without TAIT, unless we do boxing?

Was the decision made in stageless to just do the boxing?

**Comment by alice-i-cecile** [comment]
Oh right, I remember this discussion now. Sorry for the confusion. Still blocked unfortunately.

**Comment by richchurcher** [comment]
And, sadly still blocked! Closing as part of backlog cleanup. Feels like the `.chain()` syntax is fairly stable now, I guess? Last comment on the tracking thread was Dec '23, apparently [there are still issues](https://github.com/rust-lang/rust/issues/63063#issuecomment-1845797689).

**Pull request closed without merging**
